test/test_prototype_builtin_datasets.py

```diff
@@ -109,19 +110,31 @@ def test_transformable(self, test_home, dataset_mock, config):
 
         next(iter(dataset.map(transforms.Identity())))
 
-    @pytest.mark.xfail(reason="See https://github.com/pytorch/data/issues/237")
     @parametrize_dataset_mocks(DATASET_MOCKS)
-    def test_serializable(self, test_home, dataset_mock, config):
-        dataset_mock.prepare(test_home, config)
+    def test_serializable_pickle(self, mocker, test_home, dataset_mock, config):
+        if DILL_AVAILABLE:
+            mocker.patch("torch.utils.data.datapipes.datapipe.DILL_AVAILABLE", new=False)
```

Comment:
@ejguan @NivekT since the serialization backend is automatically selected based on the presence of `dill`, it is impossible to test pickle serialization without patching this. For now we only need to patch a single module, but as https://github.com/pytorch/pytorch/pull/74958#issuecomment-1084637227 implies, we need to do this in multiple places in the future. 

Would it be possible give users the option to set the serialization backend? The default can still be `dill` if available otherwise `pickle`.

Reply:
I see. I can confirm that we will rely on this `DILL_AVAILABLE` for any place in TorchData project to determine if `dill` is available or not.

> Would it be possible give users the option to set the serialization backend? The default can still be dill if available otherwise pickle.

It's doable, but I am not sure if we want to do so because the goal of automatically using `dill` is to reduce the work users need to figure out if the DataPipe is serializable with lambda function.

Reply:
> I see. I can confirm that we will rely on this `DILL_AVAILABLE` for any place in TorchData project to determine if `dill` is available or not.

The problem is that we will need to patch every single module where this is imported. You cannot patch the place where it is defined, but rather where it is used. If you look above, we are not patching `._utils.serialization` but rather `.datapipes.datapipe`, because this is where the flag is used. If we now need use this flag in multiple modules, we need to patch all of them. This is very brittle.

> It's doable, but I am not sure if we want to do so because the goal of automatically using `dill` is to reduce the work users need to figure out if the DataPipe is serializable with lambda function.

Not sure I understand. If we just keep the same detection as we have now, users that don't care should not see any difference. If `dill` is available, it will be picked up and otherwise `pickle` will be used. But it would give users the option to enforce a particular backend if they need to. Without this option, the environment you use has an effect on the functionality and there is no way change that. I don't think this is good design. 

Even if you don't do it for the users, think about how you want to test `pickle` vs `dill` yourself. Right now the only option is to have two separate workflows one with `dill` installed and one without.

Reply:
> The problem is that we will need to patch every single module where this is imported. You cannot patch the place where it is defined, but rather where it is used. If you look above, we are not patching `._utils.serialization` but rather `.datapipes.datapipe`, because this is where the flag is used. If we now need use this flag in multiple modules, we need to patch all of them. This is very brittle.

You still can add following code to override the method rather than using `patch`.
```
def state_fn(self):
    return self.__dict__
IterDataPipe. set_getstate_hook(state_fn)
```

> But it would give users the option to enforce a particular backend if they need to. Without this option, the environment you use has an effect on the functionality and there is no way change that. I don't think this is good design.

This is actually a good argument for users who have fully understanding about what they want to achieve. We may be able to expose an API to switch backend if needed, similar to the `set_getstate_hook` but with syntax sugar.

Reply:
Adding an issue https://github.com/pytorch/data/issues/341

Reply:
cc: @NivekT Overriding `set_getstate_hook` with the above function won't actually work for all DataPipe like `Forker` https://github.com/pytorch/pytorch/blob/835cc66e5dd26db558931b4fe47b45e08a3a09f7/torch/utils/data/datapipes/iter/combining.py#L158-L167

Then, we should definitely support backend switch that Philip suggested